data_safe_haven/infrastructure/components/composite/entra_application.py

```diff
@@ -0,0 +1,158 @@
+        msgraph_service_principal: Input[entra.ServicePrincipal],
+    ):
+        super().__init__(
+            application_name=application_name,
+            application_kwargs={
+                "public_client": entra.ApplicationPublicClientArgs(
+                    redirect_uris=["urn:ietf:wg:oauth:2.0:oob"]
+                )
+            },
+            application_permissions=application_permissions,
+            msgraph_service_principal=msgraph_service_principal,
+        )
+
+
+class EntraWebApplicationProps(EntraApplicationProps):
+    """
+    Properties for a web EntraApplicationComponent.
+    See https://learn.microsoft.com/en-us/entra/identity-platform/msal-client-applications)
+    """
+
+    def __init__(
+        self,
+        application_name: Input[str],
+        application_permissions: list[tuple[EntraAppPermissionType, str]],
+        msgraph_service_principal: Input[entra.ServicePrincipal],
+        redirect_url: Input[str],
+    ):
+        super().__init__(
+            application_name=application_name,
+            application_kwargs={
+                "web": entra.ApplicationWebArgs(
+                    redirect_uris=[redirect_url],
+                    implicit_grant=entra.ApplicationWebImplicitGrantArgs(
+                        id_token_issuance_enabled=True,
+                    ),
+                )
+            },
+            application_permissions=application_permissions,
+            msgraph_service_principal=msgraph_service_principal,
+        )
+
+
+class EntraApplicationComponent(ComponentResource):
+    """Deploy an Entra application with Pulumi"""
+
+    def __init__(
+        self,
+        name: str,
+        props: EntraApplicationProps,
+        opts: ResourceOptions | None = None,
+    ) -> None:
+        super().__init__("dsh:common:EntraApplicationComponent", name, {}, opts)
+
+        # Create the application
+        self.application = entra.Application(
+            f"{self._name}_application",
+            display_name=props.application_name,
+            prevent_duplicate_names=True,
+            required_resource_accesses=(
+                [
+                    entra.ApplicationRequiredResourceAccessArgs(
+                        resource_accesses=[
+                            entra.ApplicationRequiredResourceAccessResourceAccessArgs(
+                                id=props.msgraph_permissions[permission_type][
+                                    permission
+                                ],
+                                type=permission_type.value,
+                            )
+                            for permission_type, permission in props.application_permissions
+                        ],
+                        resource_app_id=props.msgraph_client_id,
+                    )
+                ]
+                if props.application_permissions
+                else []
+            ),
+            sign_in_audience="AzureADMyOrg",
```

Comment:
Is there a object in `pulumi-azuread` for this value? Is it worth making one ourselves (I think we only use it once so not a huge deal)?

Reply:
There's not, but I've added an Enum for this in b0dbcf685.